src/trackers/helpers/readers.js

```diff
@@ -0,0 +1,97 @@
+const fs = require('fs')
+const {gunzip} = require('zlib')
+const {Client} = require('pg')
+const Cursor = require('pg-cursor')
+const {promisify} = require('util')
+
+const dogunzip = promisify(gunzip)
+
+class JSONFileDataReader {
+    constructor(path) {
+        this.path = path
+        this._fileList = null
+    }
+
+    async getFileList() {
+        if (!this._fileList) {
+            this._fileList = await fs.promises.readdir(this.path)
+        }
+        return this._fileList
+    }
+
+    async length() {
+        return (await this.getFileList()).length - 1
+    }
+
+    async *iterator() {
+        for (const file of await this.getFileList()) {
+            // transparently expand gzipped files
+            const fileData = file.endsWith('.gz')
+                ? (await dogunzip(await fs.promises.readFile(`${this.path}/${file}`))).toString('utf-8')
+                : fs.promises.readFile(`${this.path}/${file}`, 'utf8')
+            const siteData = JSON.parse(fileData)
+            if (siteData.initialUrl) {
+                yield siteData
+            }
+        }
+    }
+
+    async close() {}
+}
+
+class PostgresDataReader {
+    constructor(crawlId, region) {
+        this.crawlId = crawlId
+        this.region = region
+        this._client = new Client()
+        this._connected = this._client.connect()
+    }
+
+    async close() {
+        await this._client.end()
+    }
+
+    async length() {
+        await this._connected
+        const res = await this._client.query(
+            'SELECT COUNT(*) FROM crawler WHERE crawl_id = $1 and region = $2',
+            [this.crawlId, this.region]
+        )
+        return parseInt(res.rows[0].count, 10)
+    }
+
+    async *iterator() {
+        await this._connected
+        const query = `
+        SELECT
+            initialUrl AS "initialUrl",
```

Comment:
do we need a rename if we are not changing the name? (same bellow)

Reply:
Column names in postgres are case insensitive, this means that by default the data returned would have the column name lowercased. This is a trick so that the node driver will use the literal camelCase column name, and thus return a row object with `initialUrl` as the key. The avoids having to write something like `row.initialUrl = row.initialurl` later to rename these object keys.

Reply:
Ah, TIL!